fixes #3827 - ldap avatar support

stbenjam opened the pull request:
I know this is one of the most inconsequential things ever.  It pulls the jpegPhoto attribute from LDAP as a user avatar.

~~Currently store it in the DB... (terrible, I know).  My avatar was absolutely tiny at only 4kb.  I could write to public/avatars/<hash>.jpg too, that's probably more sensible?~~

Avatars are written to /public/avatars/[hash of image].jpg.  Changes are captured if the data in LDAP changes -- I delete the old avatar and add the new one.

This capturing of LDAP changes has the added benefit of also capturing if the other attributes are changed (first name, last name, mail, etc) for a user.

lzap commented on app/helpers/home_helper.rb:
```diff
@@ -44,7 +44,7 @@ def org_switcher_title
 
   def user_header
-    summary = gravatar_image_tag(User.current.mail, :class=>'gravatar small', :alt=>_('Change your avatar at gravatar.com')) +
+    summary = avatar_image_tag(User.current, :class=>'avatar small') +
```
First - thanks for your contribution. This is great.

Mind keeping the help string if gravatar is being used?

stbenjam replied:
Ok, did this here: https://github.com/theforeman/foreman/pull/1087/files#diff-127a850b7d7b20173ad3fba88f8a25e3R304

lzap replied:
Both links goes to nowhere for me, can you rebase your work? Thanks.

stbenjam replied:
What will rebasing do? The links go to the current PR, the links work for me.  

Although I changed a lot so maybe everything needs a look, but the two relevant links went to:

 The avatar fix begins L304 in app/helpers/application_helper.rb
 Updating attributes begins L118 in  app/models/user.rb

Updating attributes probably needs to change.  Maybe only select firstname/email/lastname/avatar instead of using except?

It looks like there was some talk about this in #986 -- since my code is specific only to ldap, maybe I'm stepping on toes there.

lzap commented on app/models/auth_sources/auth_source_ldap.rb:
```diff
@@ -124,6 +129,19 @@ def required_attributes_values entry
+      [name, value.to_s]
+    end]
+  end
+ 
```
I'd appreciate a setting that would limit maximum size of a photo (before base64 encoding) with a reasonable value as a default (32 kB?)

lzap replied:
I mean as a Foreman setting option, something like ldap_avatar_max_size. I can imagine enterprises with huge 8 MB 10mbpx photos in LDAP. :-D

lzap replied:
Or maybe we can hardcode the limit as the length of the text field in the SQL database and no option would be necessary. In that case, if it exceeds the limit, I'd just throw the image away and issue a warning to the log about that. Maybe to spawn a bubble informing user that picture was too big.

lzap commented on app/models/auth_sources/auth_source_ldap.rb:
```diff
@@ -142,7 +160,7 @@ def search_for_user_entries(login, password)
                                     :filter     => object_filter & login_filter,
                                     # only ask for the DN if on-the-fly registration is disabled
-                                    :attributes => required_ldap_attributes.values)
+                                    :attributes => required_ldap_attributes.values + optional_ldap_attributes.values)
```
What happens if a user changes her photo in LDAP? I assume it is not being updated in Foreman DB now. Do we want to provide an update button?

stbenjam replied:
Good point.   I now do this here: https://github.com/theforeman/foreman/pull/1087/files#diff-4676c008b11a5480d73d4a6de01e45b9R122

Added benefit of capturing firstname/lastname/mail address changes in LDAP, too.

lzap commented on db/migrate/20131212125626_add_photo_support_to_ldap.rb:
```diff
@@ -0,0 +1,11 @@
+   remove_column :users, :photo
+   remove_column :auth_sources, :attr_photo
+  end
+end
```
I'd definitely not recommend to rely on default text length here because for Postgresql 9 and SQlite3 it maps to SQL type TEXT which is unlimited, by in MySQL it maps to TINYTEXT, TEXT, MEDIUMTEXT, or LONGTEXT2 depending on the length given (defaults to 64 kB). Not sure about sqlite3 at all, so I'd set some reasonable maximum size - let's say its 65536 and verify against then when storing base64 encoded picture.

Which might be also maximum limit of a picture LDAP size for the Foreman setting option - it is able to calculate that as base64 is predictable encoding (I think it adds 33.3% to the original data). So we can limit the setting to that.

stbenjam committed:
fixes #3827 - adds ldap avatar support

dLobatog commented:
@stbenjam Hi there, mind a rebase? :)

stbenjam commented:
@eLobato Rebased.  I suspect this will need an SELinux policy too, to ensure /public/avatars gets the right label.  

Is there an easy way to run the dev environment confined? Or build up a full set of RPM's based on a branch?

domcleal commented:
@stbenjam once it's built, click through to the sqlite3/1.9.3 icon on the build (don't ask..), and you should see a link to a build of the packaging matrix rpm job.  Click that, click the OS you want (e.g. el6) and then view the console for that job and in there you should find a link to the koji job containing RPMs for this branch :)

lzap commented:
You do not need to worry about SELinux in this case, the parent directory has the correct context, therefore all files created in that directory inherits. This is the same for the avatars subdirectory:

```
[root@hp-dl160gen8-01 ~]# touch /usr/share/foreman/public/aaa
[root@hp-dl160gen8-01 ~]# ll /usr/share/foreman/public/ -Z
-rw-r--r--. foreman foreman system_u:object_r:foreman_lib_t:s0 404.html
-rw-r--r--. foreman foreman system_u:object_r:foreman_lib_t:s0 422.html
-rw-r--r--. foreman foreman system_u:object_r:foreman_lib_t:s0 500.html
-rw-r--r--. root    root    unconfined_u:object_r:foreman_lib_t:s0 aaa
drwxr-xr-x. foreman foreman system_u:object_r:foreman_lib_t:s0 assets
-rw-r--r--. foreman foreman system_u:object_r:foreman_lib_t:s0 blank.html
-rw-r--r--. foreman foreman system_u:object_r:foreman_lib_t:s0 robots.txt
```

(I was root that is the only difference).

Sorry I missed your comment about rebase above. What is the status again?

domcleal commented:
@lzap does Foreman under SELinux (as passenger_t) even have rights to write to that dir?  I'd kind of hope not to be honest..

stbenjam commented:
@domcleal Yea, my main concern was writing to the directory.  I haven't had a chance to look at building the RPMs yet, what's the output of that packaging_build_matrix_rpm job you mentioned? 

@lzap Re: the rebase.  I rebased yesterday due to changes in develop, that's all.

The code here could use another look to make sure it looks ok, there were some small changes since last time.

domcleal commented:
The latest EL6 RPMs for your branch are here: http://koji.katello.org/koji/taskinfo?taskID=74171

lzap commented:
@stbenjam if you want, I can do the selinux package changes so you can only test this - I dont have LDAP with images here...

stbenjam commented:
@lzap foreman-selinux was easy enough to build on my own.

After some wrestling, got the RPMs from my branch installed along with foreman-selinux, and it worked, passenger can write to Rails.public_path with no policy change.

I did notice a problem, though.  path_to_image() only resolves dynamically to an assets path.  So, for this to work if foreman's running under a subdirectory, I moved /images/avatars to assets/avatars...

lzap commented:
Hmmm I wonder why do we have this, @domcleal any clue?

```
    read_files_pattern(httpd_t, foreman_lib_t, foreman_lib_t)
    manage_files_pattern(passenger_t, foreman_lib_t, foreman_lib_t)
    manage_dirs_pattern(passenger_t, foreman_lib_t, foreman_lib_t)
```

Isn't this too much?

stbenjam commented:
Passenger probably needs to modify other files in /var/lib/foreman, no?  Rails.public_path should probably have it's own context foreman_content_t.

lzap commented:
But this is in /usr/share/foreman, I think we should introduce new domain for content in /usr/share and use that. I will raise a ticket. Web app should not be able to write to directories where the application itself is located.

http://projects.theforeman.org/issues/4113

lzap commented:
I found some issues I will work on once this patch is merged. If this is working, it is fine. But you helped us to find a bug in SELinux :-)

stbenjam commented:
@lzap On el6 at least, /usr/share/foreman/public is a symlink to /var/lib/foreman/public

lzap commented:
Yeah, but this does not change the fact we should not allow to write
there. I still want to fix this.

LZ

On Fri, Jan 17, 2014 at 12:33:53AM -0800, Stephen Benjamin wrote:

> @lzap On el6 at least, /usr/share/foreman/public is a symlink to /var/lib/foreman/public
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/foreman/pull/1087#issuecomment-32587886

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

stbenjam commented:
@lzap Of course, just saying when you make the exception for avatars, to use /var/lib/foreman/public/ instead of /usr/share.

lzap commented:
Sure thing, thank you! Nice patch and testing with SELinux. stbenjam++

lzap commented:
@domcleal +1 from my side, is this blocked somehow?

domcleal commented on app/models/user.rb:
```diff
@@ -157,8 +157,18 @@ def self.try_to_login(login, password)
+        # update with returned attrs, maybe some info changed in LDAP, and clean up old avatar
+        old_hash = user.avatar_hash
+        User.as :admin do
+          user.update_attributes(attrs.slice(:firstname, :lastname, :mail, :avatar_hash)) if attrs.is_a? Hash
```
Nitpick, move the `if ..` to end of the whole User.as block (`end if attrs...`)

domcleal commented on app/models/user.rb:
```diff
@@ -157,8 +157,18 @@ def self.try_to_login(login, password)
+        User.as :admin do
+          user.update_attributes(attrs.slice(:firstname, :lastname, :mail, :avatar_hash)) if attrs.is_a? Hash
+        end
+        if user.avatar_hash != old_hash and not User.where(:avatar_hash => old_hash).any?
```
Nitpick, s/and/&&/ is preferable.  Also check `old_hash.present?` as one of the first conditions.

domcleal replied:
`User.unscoped.where(..)` too

domcleal commented:
@lzap I had no LDAP server to hand until now, and nobody else had tested it.  Now tested, works well!  Couple of minor comments above, otherwise good.

stbenjam commented:
@lzap @domcleal 

Yay! \o/ 

Thanks for taking the time to test this!!

domcleal commented:
Merged as d424cab for Foreman 1.5.0, thanks @stbenjam!
